src/tree.c

```diff
@@ -81,6 +81,30 @@ int git_tree_entry_icmp(const git_tree_entry *e1, const git_tree_entry *e2)
 		git__strncasecmp);
 }
 
+/**
+ * Allocate a tree entry, borrowing the filename from the tree which
+ * owns it. This is useful when reading trees, so we don't allocate a
+ * ton of small strings but can use the pool.
+ */
+static git_tree_entry *alloc_entry_pooled(git_pool *pool, const char *filename, size_t filename_len)
+{
+	git_tree_entry *entry = NULL;
+	size_t tree_len;
+
```

Comment:
Absolutely, I've pushed up a commit which does this and de-duplicates the size and overflow checks.